run-tests.php

```diff
@@ -1209,6 +1268,299 @@ function run_all_tests($test_files, $env, $redir_tested = null)
+	// parallel execution in mind, it is not safe to just run any two tests
+	// concurrently.
+	// Therefore, we divide the test set into directories and test multiple
+	// directories at once, but not multiple tests within them.
+
+	$testDirs = [];
+
+	foreach ($test_files as $file) {
+		$dirSeparator = strrpos($file, DIRECTORY_SEPARATOR);
+		if ($dirSeparator !== FALSE) {
+			$testDirs[substr($file, 0, $dirSeparator)][] = $file;
+		} else {
+			$testDirs[""][] = $file;
+		}
+	}
+
+	$testDirsToGo = array_values($testDirs);
+	// Sort test dirs so the biggest ones are handled first, so we spend less
+	// time waiting on workers tasked with very large dirs.
+	// This is an ascending sort because items are popped off the end.
+	// Thank you Rasmus for this idea :)
+	usort($testDirsToGo, function ($a, $b) {
+		return count($a) - count($b);
+	});
+	$testDirsInProgress = 0;
+
+	echo "Isolated ", count($testDirsToGo), " directories to be tested in parallel.\n";
+
+	echo "Spawning workers… ";
+	for ($i = 1; $i <= $workers; $i++) {
+		$proc = proc_open(
+			$thisPHP . ' ' . escapeshellarg($thisScript),
+			[
+				0 => ['pipe', 'r'],
+				1 => ['pipe', 'w'],
+				2 => ['pipe', 'w']
+			],
+			$pipes,
+			NULL,
+			$_ENV + [
+				"TEST_PHP_WORKER" => $i
+			],
+			[
+				"suppress_errors" => TRUE
+			]
+		);
+		if ($proc === FALSE) {
+			kill_children($workerProcs);
+			error("Failed to spawn worker $i");
+		}
+		$workerProcs[$i] = $proc;
+
+		$greeting = base64_encode(serialize([
+			"type" => "hello",
+			"workerID" => $i,
+			"GLOBALS" => $GLOBALS,
+			"constants" => [
+				"INIT_DIR" => INIT_DIR,
+				"TEST_PHP_SRCDIR" => TEST_PHP_SRCDIR,
+				"PHP_QA_EMAIL" => PHP_QA_EMAIL,
+				"QA_SUBMISSION_PAGE" => QA_SUBMISSION_PAGE,
+				"QA_REPORTS_PAGE" => QA_REPORTS_PAGE,
+				"TRAVIS_CI" => TRAVIS_CI
+			]
+		])) . "\n";
+
+		stream_set_timeout($pipes[0], 5);
+		if (fwrite($pipes[0], $greeting) === FALSE) {
+			kill_children($workerProcs);
+			error("Failed to send greeting to worker $i.");
+		}
+
+		stream_set_timeout($pipes[1], 5);
+		$rawReply = fgets($pipes[1]);
+		if ($rawReply === FALSE) {
+			kill_children($workerProcs);
+			error("Failed to read greeting reply from worker $i.");
+		}
+
+		$reply = unserialize(base64_decode($rawReply));
+		if (!$reply || $reply["type"] !== "hello_reply" || $reply["workerID"] !== $i) {
+			kill_children($workerProcs);
+			error("Greeting reply from worker $i unexpected or could not be decoded: '$rawReply'");
+		}
+
+		$workerStdins[$i] = $pipes[0];
+		$workerStdouts[$i] = $pipes[1];
+		$workerStderrs[$i] = $pipes[2];
+
+		echo "$i ";
+	}
+	echo "… done!\n";
+	echo "====⚡️===========================================================⚡️====\n";
+	echo "\n";
+
+escape:
+	while ($testDirsToGo || ($testDirsInProgress > 0)) {
+		$toRead = array_values($workerStdouts);
+		$toWrite = NULL;
+		$toExcept = NULL;
+		if (stream_select($toRead, $toWrite, $toExcept, 0, 50 * 1000)) {
```

Comment:
@kelunik I'd guess you know what you're doing here better than I do, so: Would you see any disadvantage to having a higher timeout here, say 1 second? Checking the reads on these is supposedly non-blocking, and there's nothing else the host process needs to do except for wait on reads, so reducing the number of syscalls very slightly would make sense, right?

Reply:
Yes, you can totally increase that to 10 seconds or even more. It will be interrupted by any signals or actionable streams.